Source/CarthageKit/Project.swift

```diff
@@ -689,50 +710,80 @@ public final class Project {
-		let rawDependencyURL = rootDirectoryURL.appendingPathComponent(dependency.relativePath, isDirectory: true)
+	private func symlinkCheckoutPathsForDependencyProject(dependency: Dependency<PinnedVersion>, repositoryURL: URL, rootDirectoryURL: URL, addSubmoduleForDependency: SignalProducer<(), CarthageError>? = nil) -> SignalProducer<(), CarthageError> {
+		let rawDependencyURL = rootDirectoryURL.appendingPathComponent(dependency.project.relativePath, isDirectory: true)
 		let dependencyURL = rawDependencyURL.resolvingSymlinksInPath()
 		let dependencyCheckoutsURL = dependencyURL.appendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true).resolvingSymlinksInPath()
-		let subDependencyNames = subDependencies.map { $0.name }
-		let fileManager = FileManager.`default`
+		let fileManager = NSFileManager.`default`
+
+		return self.dependencyProjects(for: dependency)
+			.zip(with: // file system objects from git in `CarthageProjectCheckoutsPath` which might conflict with symlinks.
+				launchGitTask(
+					// ls-tree, because `ls-files` returns no output (for all instances I've seen) on bare repos.
+					// flag “-z” enables output separated by the nul character (`\0`).
+					[ "ls-tree", "-z", "-r", "--full-name", "--name-only", dependency.version.commitish, CarthageProjectCheckoutsPath ],
+					repositoryFileURL: repositoryURL
+				)
+					.map { (output: String) -> [String] in
+						output.characters.lazy
+							.split("\0", allowEmptySlices: false)
+							.map { String($0) }
+							.flatMap { (path: String) -> String? in
+								let componentsRelativeToDirectoryURL = {
+									return URL(string: $0, relativeToURL: self.directoryURL)?.standardizedFileURL.carthage_pathComponents
+								}
+								if
+									let components = componentsRelativeToDirectoryURL(path),
+									let comparator = componentsRelativeToDirectoryURL(CarthageProjectCheckoutsPath)
+									where Array(components.dropLast(1)) == comparator
+								{
+									return components.last!
+								} else {
+									return nil
+								}
+							}
+					}
+			)
+			.attemptMap { (dependencies: Set<ProjectIdentifier>, components: [String]) -> Result<(), CarthageError> in
+				let names = dependencies.filter { dependency in
+					// Filter out dependencies with names matching (case-insensitively) file system objects from git in `CarthageProjectCheckoutsPath`.
+					// Edge case warning on file system case-sensitivity. If a differently-cased file system object exists in git
+					// and is stored on a case-sensitive file system (like the Sierra preview of APFS), we currently preempt
+					// the non-conflicting symlink. Probably, nobody actually desires or needs the opposite behavior.
+					components.filter {
+						dependency.name.caseInsensitiveCompare($0) == .orderedSame
+					}.isEmpty
+				}.map { $0.name }
+
+				if let result = addSubmoduleForDependency?.wait(), case .Failure(_) = result { return result }
```

Comment:
We really don't want to use `wait`. The only wait should be at the top-most level.